src/test/java/net/kaczmarzyk/spring/data/jpa/utils/SpecificationBuilderTest.java

```diff
@@ -0,0 +1,118 @@
+import net.kaczmarzyk.spring.data.jpa.web.SpecificationFactory;
+import net.kaczmarzyk.spring.data.jpa.web.annotation.Join;
+import net.kaczmarzyk.spring.data.jpa.web.annotation.Or;
+import net.kaczmarzyk.spring.data.jpa.web.annotation.Spec;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.springframework.data.jpa.domain.Specification;
+
+import javax.persistence.criteria.JoinType;
+import java.util.List;
+
+import static net.kaczmarzyk.spring.data.jpa.utils.SpecificationBuilder.specification;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
+
+/**
+ * @author Jakub Radlica
+ * @author Kacper Leśniak (Tratif sp. z o.o.)
+ */
+public class SpecificationBuilderTest extends E2eTestBase {
+
+	@Join(path = "orders", alias = "o")
+	@Join(path = "o.tags", alias = "t", type = JoinType.INNER)
+	@Or({
+			@Spec(path = "o.itemName", pathVars = "orderIn", spec = In.class),
+			@Spec(path = "t.name", headers = "tag", spec = Equal.class)
+	})
+	public interface CustomSpecificationWithPathVar extends Specification<Customer> {
+	}
+
+	@Join(path = "orders", alias = "o")
+	@Join(path = "o.tags", alias = "t", type = JoinType.INNER)
+	@Or({
+			@Spec(path = "o.itemName", params = "orderIn", spec = In.class),
+			@Spec(path = "t.name", headers = "tag", spec = Equal.class)
+	})
+	public interface CustomSpecificationWithParam extends Specification<Customer> {
+	}
+
+	@Join(path = "orders", alias = "o")
+	@Join(path = "o.tags", alias = "t", type = JoinType.INNER)
+	@Or({
+			@Spec(path = "o.itemName", headers = "orderIn", spec = In.class),
+			@Spec(path = "t.name", params = "tag", spec = Equal.class)
+	})
+	public interface CustomSpecificationWithHeader extends Specification<Customer> {
+	}
+
+	@Test
+	public void shouldCreateSpecificationDependingOnPathVar() {
+		Specification<Customer> spec = specification(CustomSpecificationWithPathVar.class)
+				.withPathVar("orderIn", "Pizza")
+				.build();
+
+		List<Customer> customers = customerRepo.findAll(spec);
+
+		assertThat(customers.size()).isEqualTo(1);
+		assertThat(customers.get(0).getFirstName()).isEqualTo("Homer");
+	}
+
+	@Test
+	public void shouldCreateSpecificationDependingOnParam() {
+		Specification<Customer> spec = specification(CustomSpecificationWithParam.class)
+				.withParam("orderIn", "Pizza")
+				.build();
+
+		List<Customer> customers = customerRepo.findAll(spec);
+
+		assertThat(customers.size()).isEqualTo(1);
+		assertThat(customers.get(0).getFirstName()).isEqualTo("Homer");
+	}
+
+	@Test
+	public void shouldCreateSpecificationDependingOnHeader() {
+		Specification<Customer> spec = specification(CustomSpecificationWithHeader.class)
+				.withHeader("orderIn", "Pizza")
+				.build();
+
+		List<Customer> customers = customerRepo.findAll(spec);
+
+		assertThat(customers.size()).isEqualTo(1);
+		assertThat(customers.get(0).getFirstName()).isEqualTo("Homer");
+	}
+
+	@Test
+	public void shouldUseCreateSpecificationDependingOnMethod() {
```

Comment:
"should use create specification ..." -- I don't understand this sentence, please rename